### PR TITLE
perf: optimize the modification of rollout to httproute header

### PR DESCRIPTION
**Pull request by ZhangSetSail**
### Ⅰ. Describe what this PR does
rollout 对 httproute 的修改存在缺陷，面对 header 的或逻辑并没有支持。

### Ⅱ. Does this pull request fix one issue?

是的

### CODE REVIEW

**Comment by ZhangSetSail** [comment]
https://github.com/openkruise/rollouts/issues/138

**Comment by ZhangSetSail** [comment]
还存在问题，我再看看

**Comment by ZhangSetSail** [comment]
测试没问题了，镜像地址：registry.cn-hangzhou.aliyuncs.com/zhangqihang/kruise-rollout:v0428-1

**Comment by ZhangSetSail** [comment]
这个问题放置好久了哇，我还需要做些什么嘛

**Comment by furykerry** [comment]
> 测试没问题了，镜像地址：registry.cn-hangzhou.aliyuncs.com/zhangqihang/kruise-rollout:v0428-1

sorry for the delay, we miss the PR, the PR looks generally good, but i don't have time to review it yet. Anyway, you can rebase the master to make the e2e pass

File: pkg/trafficrouting/network/gateway/gateway.go [hunk 1]
```diff
@@ -159,10 +159,16 @@ func (r *gatewayController) buildCanaryHeaderHttpRoutes(rules []gatewayv1alpha2.
 		canaryRef.Name = gatewayv1alpha2.ObjectName(r.conf.CanaryService)
 		canaryRule.BackendRefs = []gatewayv1alpha2.HTTPBackendRef{*canaryRef}
 		// set canary headers in httpRoute
+		var newMatches []gatewayv1alpha2.HTTPRouteMatch
 		for j := range canaryRule.Matches {
-			match := &canaryRule.Matches[j]
-			match.Headers = append(match.Headers, headers...)
+			canaryRuleMatch := &canaryRule.Matches[j]
+			for k := range matchs {
+				match := canaryRuleMatch.DeepCopy()
```
Comment (furykerry) [inline pkg/trafficrouting/network/gateway/gateway.go:166]: it seems not necessary to deepcopy here since canaryRule is already a deep clone of rule.

Comment (ZhangSetSail) [reply pkg/trafficrouting/network/gateway/gateway.go:166]: 好的，我修改一下。

**Comment by ZhangSetSail** [comment]
The inspection has not been passed, this pr is too messy, I try to submit a new pr, this one is closed https://github.com/openkruise/rollouts/pull/154

File: pkg/trafficrouting/network/gateway/gateway.go (hunk 1 above)
Comment (ZhangSetSail) [reply pkg/trafficrouting/network/gateway/gateway.go]: I tried to delete the deep copy, but it seems that it is not allowed, and there will be problems with the obtained data.
expect:
```
 [
    {
        "matches":[
            {
                "path":{
                    "value":"/web"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"web-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/list"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"list-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/storage"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc-canary",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/storage"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/storage"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc-canary",
                "port":8080
            }
        ]
    }
]
```
but get:
```
[
    {
        "matches":[
            {
                "path":{
                    "value":"/web"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"web-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/list"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"list-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/storage"
                }
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/store"
                },
                "headers":[
                    {
                        "name":"version",
                        "value":"v2"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/v2/store"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc-canary",
                "port":8080
            }
        ]
    },
    {
        "matches":[
            {
                "path":{
                    "value":"/storage"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            },
            {
                "path":{
                    "value":"/storage"
                },
                "headers":[
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"123*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    },
                    {
                        "type":"RegularExpression",
                        "name":"user_id",
                        "value":"234*"
                    },
                    {
                        "name":"canary",
                        "value":"true"
                    }
                ]
            }
        ],
        "backendRefs":[
            {
                "kind":"Service",
                "name":"store-svc-canary",
                "port":8080
            }
        ]
    }
]
```

**Pull request merged**